### PR TITLE
[tosa] aten.embedding op support

### PR DESCRIPTION
**Pull request by sjarus**
Enables BERT legalization.

Signed-off-by: Suraj Sudhir <suraj.sudhir@arm.com>

### CODE REVIEW

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 1]
```diff
@@ -2609,6 +2609,80 @@ LogicalResult ConvertAtenOp<AtenGeluBackwardOp>::matchAndRewrite(
...
+    return op.emitError("Indices must be of integer tensor type");
+
+  if (indicesType.getRank() != 2)
+    return op.emitError("indices must be of rank 2");
+
+  auto weightType = weight.getType().cast<RankedTensorType>();
+  if (weightType.getRank() != 2)
+    return op.emitError("weight must be of rank 2");
+
+  // FIXME: padding_idx, scale_grad_by_freq and sparse are not handled yet.
```
Comment (silvasean) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:2633]: can you make sure to bail out on those cases?

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 2]
```diff
@@ -3173,11 +3247,11 @@ class ConvertTorchToTosa : public ConvertTorchToTosaBase<ConvertTorchToTosa> {
                                            tosa::AvgPool2dOp);
 #undef INSERT_ADAPTIVE_POOLING_ATEMOP_PATTERN
 
-target.addIllegalOp<AtenMaxPool2dOp>();
-patterns.add<ConvertAtenMaxPool2dOp>(typeConverter, context);
+    target.addIllegalOp<AtenMaxPool2dOp>();
```
Comment (silvasean) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:3250]: nit: why the indent?

**Review by silvasean — APPROVED** [review]

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 2 above)
Comment (sjarus) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3250]: Yeah that's weird. Looks like a preexisting condition that's not covered by my insurance here though. I'm surprised the clang-format didn't fix it, or maybe it's what broke it.

**Comment by nithinsubbiah** [comment]
Hi @sjarus, I tried out this patch to lower minilm to TOSA but it fails with a seg fault specifically an assertion error.
```
torch-mlir/externals/llvm-project/llvm/include/llvm/ADT/SmallVector.h:273: llvm::SmallVectorTemplateCommon::reference llvm::SmallVectorTemplateCommon<TensorShape_t>::operator[](llv
m::SmallVectorTemplateCommon::size_type) [T = TensorShape_t]: Assertion `idx < size()' failed.
```
The script that I used is here (https://gist.github.com/nithinsubbiah/66c3a9137f254f5d327929cfef11d879), I was able to get Torch IR with no dynamic shapes.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 2 above)
Comment (silvasean) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:3250]: Don't worry about it. Yeah, fine to leave it.

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp (hunk 1 above)
Comment (sjarus) [reply lib/Conversion/TorchToTosa/TorchToTosa.cpp:2633]: Done

**Review by silvasean — APPROVED** [review]

**Pull request merged**